### ISSUE TITLE
Equations: multiply by x on both sides

### ISSUE THREAD

**Issue by evykassirer** [socraticorg/mathsteps#51]
Blocked on #137 and #139 because without cancelling working, we can't multiply both sides by the denominator and then cancel

---

e.g. `2/x = 1 --> 2/x * x =  1 * x`

Currently there's a function for this in `EquationOperations` called `removeSymbolFromDenominator` but it's not implemented yet.

It should work like this:
- **if** the left node of an equation is a fraction
- **and** the denominator of the fraction has `symbolName` in it (so the fraction could be `1/x` or `(3 + x) / (x^2 + 2)` etc.) 
- **then** multiply both sides by the denominator

**Comment by arbylee** [comment]
Hi there, I'm interested in taking a crack at this if it's not assigned

**Comment by evykassirer** [comment]
That would be fantastic, thank you! This will help us solve a lot more equations than we were able to before.

Let me know if you have any questions or encounter anything tricky

**Comment by arbylee** [comment]
I got a basic case working, but am questioning my approach as I try to get the more complex case working. I'd like to hear your thoughts

Symbols.getLastSymbolTerm currently relies on Node.PolynomialTerm.isPolynomialTerm to find the last symbol term.  It doesn't recognize terms such as `1/(2+x)` because it's not definitionally a polynomial term. I'm not planning on changing isPolynomialTerm, but I've started to enhance Symbols.getLastSymbolTerm and the helper function isSymbolTerm to recognize `1/(2+x)` as a symbol term. I'm currently planning something like Symbols.getLastSymbolTerm on `3/(x^2 + 3)` to return `1/(x^2 + 3)` I'm wondering if that's dirtying the definition of "symbol term" or if you have some other approach in mind

**Comment by evykassirer** [comment]
Yeah I had a feeling this had some trickiness to it haha.

Good idea - I think that makes sense (and if it doesn't, we can move that functionality to a different place so feel free to write it out now). Perhaps this function shouldn't live in Symbols and should be in a more equation specific file. What do you think?

I'd also like to bring @aelnaiem into the conversation too if he's free/interested, cause he wrote most of the equation code and might have some opinions on stuff.

**Comment by sangwinc** [comment]
There are some interesting design decisions to be made here.  E.g. how will your system avoid the following (well-known) nonsense?

a  = b
a^2 = ab 
a^2 -b^2 = ab -b^2  
(a-b)(a+b) = b(a-b)  
a+b = b 
2b = b  
2 = 1 

(If you are interested in this kind of thing, you might like the attached paper)

@ARTICLE{2015Audited,
  author =       {Sangwin, C.~J.},
  title =        {An Audited Elementary Algebra},
  journal =      {The Mathematical Gazette},
  year =         {2015},
  month =        {July},
  volume =       {99},
  number =       {545},
  pages =        {290--297},
  doi =          {http://dx.doi.org/10.1017/mag.2015.37},
}

Chris
[2015-AuditedAlgebra.pdf](https://github.com/socraticorg/mathsteps/files/723419/2015-AuditedAlgebra.pdf)

**Comment by hmaurer** [comment]
@sangwinc Interesting; I guess you would need some form of "pre-condition" handling, and branching as in https://github.com/socraticorg/mathsteps/issues/49#issuecomment-274154703. This could easily get more complex though. 

I believe right now the system is purely based on trying out strategies by order of "priority" and applying the one it can, @evykassirer ? When doing elementary algebra this seems very close to what a student would do, since they have yet to develop intuition. But as they get better they'll know some branches are not worth exploring, and that doesn't seem to be something doable in the current architecture. This will show up both it equation solving and simplifications (e.g. https://github.com/socraticorg/mathsteps/issues/45#issuecomment-274156859). There would need to be a balance between the "try strategies until one works" approach and the "be clever and look a couple of steps ahead for the best strategy". Ideally this would even adapt depending on the student...

**Comment by evykassirer** [comment]
Right now we deal with this by getting to some "standard form" (using strategies by priority like @hmaurer says) and then deciding if it makes sense or not when we get there. It's easy for now because we only allow max one variable (either both sides are constants and not equivalent, in which case we return that the equation cannot hold true, or the sides are equivalent and the equation holds true, or we get something like `x=3` and there's a solution, or there's something more complicated - which we just don't handle yet)

@sangwinc can you explain a bit more about how the way we design multiplying by the denominator will affect this? or are you generalizing the conversation here to be more about the future of how we deal with equations.

Thanks for the research paper! Maybe we can make a RESOURCES.md file with relevant research and projects :)

**Comment by sangwinc** [comment]
Yes, I am generalising the conversation and suggest you at least have one eye on what is (soon) coming round the corner.

If you have polynomial equations in one variable then things are relatively simple, but you still have to worry about "loosing" a root.  E.g.   x(x-2)=0  -> x=2  and you've missed off the root x=0.  This is so obvious here that you hope student's won't do this, but what about this one?  It won't be very long until you want to include equations like this.

\begin{align*}
\frac{x+5}{x-7}-5 & = \frac{4x-40}{13-x}\\
\frac{x+5-5(x-7)}{x-7} & = \frac{4x-40}{13-x}\\
\frac{4x-40}{7-x} & = \frac{4x-40}{13-x}\\
13-x & = 7-x\\
13 & = 7.
\end{align*}
